### PR TITLE
Meson format

### PR DESCRIPTION
**Pull request by bruchar1**
This is a native `meson format` command, similar to `muon fmt`.

### CODE REVIEW

**Comment by tristan957** [comment]
cc @JCWasmx86

**Comment by bruchar1** [comment]
rebased, and targeted for 1.4.0

File: mesonbuild/ast/printer.py [hunk 1]
```diff
@@ -66,9 +66,7 @@ def visit_NumberNode(self, node: mparser.NumberNode) -> None:
         node.lineno = self.curr_line or node.lineno
 
     def escape(self, val: str) -> str:
-        return val.translate(str.maketrans(T.cast(
-            'T.Dict[str, T.Union[str, int]]',
-            {'\'': '\\\'', '\\': '\\\\'})))
+        return val.replace('\\', '\\\\').replace("'", "\'")
```
Comment (dcbaker) [inline mesonbuild/ast/printer.py:60]: I'm really shocked here, but this is about 10x faster than the `str.translate` approach. I expected the opposite.

File: docs/markdown/Commands.md [hunk 2]
```diff
@@ -396,3 +396,74 @@ format should be used. There are currently 3 formats supported:
...
+#### Differences with `muon fmt`
+
+The `meson format` command should be compatible with the `muon fmt` command.
+However, it has more features, and some diffences:
+
+- By default, `meson format` put two spaces before inline comments,
+  while `muon fmt` only puts one.
+- `muon fmt` can potentially mix crlf and lf end-of-lines, as it is not aware
+  of them. `meson format` will always be consistent in the output it produces.
+- `muon fmt` only recognize the
```
Comment (dcbaker) [inline docs/markdown/Commands.md:466]: Something is missing here

Comment (bruchar1) [reply docs/markdown/Commands.md]: Yes. Not sure what happened. It should be fixed now (along with a few other typos)

File: mesonbuild/mformat.py [hunk 3]
```diff
@@ -0,0 +1,966 @@
...
+    parser.add_argument(
+        '-r', '--recursive',
+        action='store_true',
+        help='recurse subdirs (requires --check-only or --inplace option)',
+    )
+    parser.add_argument(
+        '-c', '--configuration',
+        metavar='meson_fmt.ini',
+        type=Path,
+        help='read configuration from meson_fmt.ini'
```
Comment (dcbaker) [inline mesonbuild/mformat.py:886]: Can we call this `meson.format` instead of `meson_fmt.in`?

**Review by dcbaker — CHANGES_REQUESTED** [review]
A few of these commits could use more detailed commit messages.

The first one I suspect (though I haven't checked) was originally only for keyword arguments so column made sense, and stopped making sense as we extended it to additional places.

Overall I like this a lot, and while there's some additional things I'd like to handle here, I think those are fine follow ons. 

I t would be nice if `meson format` would automatically load a `meson.format` file that was at the root of the source directory so that proejcts could just run `meson format` and not have to worry about passing in the format rules if they're defined globally

File: docs/markdown/Commands.md [hunk 4]
```diff
@@ -396,3 +396,74 @@ format should be used. There are currently 3 formats supported:
...
+*(since 1.5.0)*
+
+{{ format_usage.inc }}
+
+Format specified `meson.build` documents. For compatibility with `muon`, `fmt`
+is an alias to `format`.
+
+{{ format_arguments.inc }}
+
+The configuration file is a `.ini` file.
```
[the remark this answers is not in the document]
Comment (bruchar1) [reply docs/markdown/Commands.md:412]: The problem I see is Python has toml support only from 3.11.

File: docs/markdown/Commands.md [hunk 5]
```diff
@@ -396,3 +396,74 @@ format should be used. There are currently 3 formats supported:
...
+    contain substitutions (default is true).
+- inserty_final_newline (bool): If true, force the `meson.build` file
+    to end with a newline character (default is true).
+- tab_width (int): Width of tab stops, used to compute line length
+    when `indent_by` uses tab characters (default is 4).
+- sort_files (bool): When true, arguments of `files()` function are
+    sorted alphabetically (default is false).
+- group_arg_value (bool): When true, string argument with `--` prefix
+    followed by string argument without `--` prefix are grouped on the
+    same line, in multiline arguments (default is false).
```
[the remark this answers is not in the document]
Comment (bruchar1) [reply docs/markdown/Commands.md:451]: Good idea! Done.

File: mesonbuild/mformat.py (hunk 3 above)
Comment (bruchar1) [reply mesonbuild/mformat.py]: Done. I also added the feature to automatically load `meson.format` when it is present and no config file is specified on the command line.

**Comment by bruchar1** [comment]
> A few of these commits could use more detailed commit messages.

I tried to improve it. But I am open to suggestions...

> I t would be nice if `meson format` would automatically load a `meson.format` file that was at the root of the source directory so that proejcts could just run `meson format` and not have to worry about passing in the format rules if they're defined globally

Done. Will also automatically analyse meson.build in current directory if it is present and no other build files are specified on the command line.

File: docs/markdown/Commands.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcbaker) [reply docs/markdown/Commands.md]: Yeah, I don’t think we can, but I can wish. Lol

**Review by dcbaker — APPROVED** [review]
Looks great, thanks!

**Pull request merged**